RecoLocalMuon/GEMSegment/plugins/GEMSegAlgo.cc

```diff
@@ -0,0 +1,479 @@
+  std::vector<GEMSegment>          segments;
+  ProtoSegments rechits_clusters; // this is a collection of groups of rechits
+  
+  if(preClustering) {
+    // run a pre-clusterer on the given rechits to split obviously separated segment seeds:
+    if(preClustering_useChaining){
+      // it uses X,Y,Z information; there are no configurable parameters used;
+      // the X, Y, Z "cuts" are just (much) wider than reasonable high pt segments
+      edm::LogVerbatim("GEMSegAlgo") << "[GEMSegAlgo::run] preClustering :: use Chaining";
+      rechits_clusters = this->chainHits( rechits );
+    }
+    else{
+      // it uses X,Y information + configurable parameters
+      edm::LogVerbatim("GEMSegAlgo") << "[GEMSegAlgo::run] Clustering";
+      rechits_clusters = this->clusterHits(rechits );
+    }
+    // loop over the found clusters:
+      edm::LogVerbatim("GEMSegAlgo") << "[GEMSegAlgo::run] Loop over clusters and build segments";
+    for(auto sub_rechits = rechits_clusters.begin(); sub_rechits !=  rechits_clusters.end(); ++sub_rechits ) {
+      // clear the buffer for the subset of segments:
+      segments_temp.clear();
+      // build the subset of segments:
+      segments_temp = this->buildSegments( (*sub_rechits) );
+      // add the found subset of segments to the collection of all segments in this chamber:
+      segments.insert( segments.end(), segments_temp.begin(), segments_temp.end() );
+    }
+  
+
+    return segments;
+  }
+  else {
+    segments = this->buildSegments(rechits);
+    return segments;
+  }
+}
+
+
+// ********************************************************************;
+GEMSegAlgo::ProtoSegments 
+GEMSegAlgo::clusterHits(const EnsembleHitContainer& rechits) {
+
+  // think how to implement BX requirement here
+
+  ProtoSegments rechits_clusters; // this is a collection of groups of rechits
+
+  float dXclus_box = 0.0;
+  float dYclus_box = 0.0;
+
+  EnsembleHitContainer temp;
+  ProtoSegments seeds;
+
+  std::vector<float> running_meanX;
+  std::vector<float> running_meanY;
+
+  std::vector<float> seed_minX;
+  std::vector<float> seed_maxX;
+  std::vector<float> seed_minY;
+  std::vector<float> seed_maxY;
+
+  // split rechits into subvectors and return vector of vectors:
+  // Loop over rechits 
+  // Create one seed per hit
+  for(unsigned int i = 0; i < rechits.size(); ++i) {
+    temp.clear();
+    temp.push_back(rechits[i]);
+    seeds.push_back(temp);
+
+    // First added hit in seed defines the mean to which the next hit is compared
+    // for this seed.
+    // running_meanX.push_back( rechits[i]->localPosition().x() );
+    // running_meanY.push_back( rechits[i]->localPosition().y() );
+    GEMDetId rhID                  = rechits[i]->gemId();
+    const GEMEtaPartition * rhEP   = (theEnsemble.second.find(rhID.rawId()))->second;
+    if(!rhEP) throw cms::Exception("GEMEtaPartition not found") << "Corresponding GEMEtaPartition to GEMDetId: "<<rhID<<" not found in the GEMEnsemble";
```

Comment:
I suppose this can only indicate there is a problem in configuration or there is a bug in the code.
If not, there should be some other way to handle the run time error.